src/compiler/crystal/semantic/math_interpreter.cr

```diff
@@ -100,31 +109,35 @@ struct Crystal::MathInterpreter
     end
 
     if visitor.expand_macro(node, raise_on_missing_const: false, first_pass: true)
-      return interpret(node.expanded.not_nil!)
+      return interpret?(node.expanded.not_nil!)
     end
 
     nil
   end
 
-  def interpret(node : Path)
+  def interpret?(node : Path)
     type = @path_lookup.lookup_type_var(node)
     case type
     when Const
-      interpret(type.value)
+      interpret?(type.value)
     else
-      node.raise "invalid constant value"
+      fail node, "invalid constant value"
     end
   end
 
-  def interpret(node : Expressions)
+  def interpret?(node : Expressions)
     if node.expressions.size == 1
-      interpret(node.expressions.first)
+      interpret?(node.expressions.first)
     else
-      node.raise "invalid constant value"
+      fail node, "invalid constant value"
     end
   end
 
+  def interpret?(node : ASTNode)
+    fail node, "invalid constant value"
+  end
+
   def interpret(node : ASTNode)
-    node.raise "invalid constant value"
+    interpret?(node) || @error.try { |(node, message)| node.raise(message) } || node.raise("invalid constant value")
```

Comment:
I don't think storing the error in a class var is a good idea. Instead, I'd suggest to avoid raising in the nilable method by returning or yielding an error instance.

Reply:
Agreed. But in this case, it has recursive calls: `interpret -> interpret_call_macro -> interpret` and blocks can't be inlined. Is returning an error/value union fine?

Reply:
Right. Returning a union should be fine.